File: src/main/java/org/smarthata/service/tm/command/LightCommand.java
package org.smarthata.service.tm.command;

import org.smarthata.service.device.LightService;
import org.springframework.stereotype.Service;
import org.telegram.telegrambots.meta.api.methods.BotApiMethod;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

@Service
public class LightCommand extends AbstractCommand {

    private static final String LIGHT = "light";
    private static final Map<String, String> translations = Map.of(
            "all", "Везде",
            "bathroom", "Ванная",
            "bedroom", "Спальня",
            "canopy", "Навес",
            "room-egor", "Детская Егора",
            "room-liza", "Детская Лизы",
            "stairs-night", "Ночник на лестнице",
            "stairs", "Свет на лестнице");

    private final LightService lightService;

    public LightCommand(LightService lightService) {
        super(LIGHT);
        this.lightService = lightService;
    }

    @Override
    public BotApiMethod<?> answer(CommandRequest request) {

        String text = "Освещение в комнатах:";

        int temporary = 0;
        if (request.hasNext()) {
            String part = request.next();
            switch (part) {
                case "1min" -> temporary = 1;
                case "5min" -> temporary = 5;
            }
            if (request.hasNext()) {
                if (temporary > 0) {
                    String room = request.next();
                    lightService.enableLightTemporary(room, TimeUnit.MINUTES.toSeconds(temporary));
                    text = "Принято " + room + "!";
                } else {
                    String action = request.next();
                    switch (action) {
                        case "on" -> {
                            lightService.setLight(part, "1");
                            text = "Включено\n" + text;
                        }
                        case "off" -> {
                            lightService.setLight(part, "0");
                            text = "Выключено\n" + text;
                        }
                    }
                }
            }
            if (temporary > 0) return showTemporaryView(request, temporary, text);
        }


        return showMainView(request, text);
    }

    private BotApiMethod<?> showMainView(CommandRequest request, String text) {
        // /light
        Map<String, String> rooms = new LinkedHashMap<>();
        lightService.getLightState().forEach((room, roomState) -> {
            String action = !roomState ? "on" : "off";
            String currentStatus = roomState ? "on" : "off";
            rooms.put(room + "/" + action,
                    translations.getOrDefault(room, room) + ": " + currentStatus);
        });

        rooms.put("1min", "1 мин");
        rooms.put("5min", "5 мин");
        rooms.put("back", "Назад");
        InlineKeyboardMarkup buttons = createButtons(List.of(), rooms, 2);
        return createTmMessage(request.getChatId(), request.getMessageId(), text, buttons);
    }

    private BotApiMethod<?> showTemporaryView(CommandRequest request, int temporary, String text) {
        // /light
        text += "\nВключить на %d мин:".formatted(temporary);

        Map<String, String> rooms = new LinkedHashMap<>();
        lightService.getLightState().forEach((room, roomState) -> {
            String currentStatus = roomState ? "on" : "off";
            rooms.put(room,
                    translations.getOrDefault(room, room) + ": " + currentStatus);
        });
        rooms.put("back", "Назад");
        List<String> path = request.getPath();
        if (path.size() > 2) path = path.subList(0, 1);
        InlineKeyboardMarkup buttons = createButtons(path, rooms, 2);
        return createTmMessage(request.getChatId(), request.getMessageId(), text, buttons);
    }

}

File: src/main/kotlin/org/smarthata/service/tm/command/LightCommand.kt
package org.smarthata.service.tm.command

import org.smarthata.service.device.LightService
import org.smarthata.service.device.Room
import org.smarthata.service.message.EndpointType.TELEGRAM
import org.springframework.stereotype.Service
import org.telegram.telegrambots.meta.api.methods.BotApiMethod
import java.util.concurrent.TimeUnit.MINUTES

@Service
class LightCommand(private val lightService: LightService) : AbstractCommand(LIGHT) {
    override fun answer(request: CommandRequest): BotApiMethod<*> {
        var text = "Освещение в комнатах:"

        var temporary = 0
        if (request.hasNext()) {
            val part = request.next()
            when (part) {
                "1min" -> temporary = 1
                "5min" -> temporary = 5
            }
            if (request.hasNext()) {
                if (temporary > 0) {
                    val room = request.next()
                    lightService.enableLightTemporary(room, MINUTES.toSeconds(temporary.toLong()), TELEGRAM)
                    text = "Принято $room!"
                } else {
                    val action = request.next()
                    when (action) {
                        "on" -> {
                            lightService.updateLight(part, true, TELEGRAM)
                            text = "Включено\n$text"
                        }

                        "off" -> {
                            lightService.updateLight(part, false, TELEGRAM)
                            text = "Выключено\n$text"
                        }
                    }
                }
            }
            if (temporary > 0) return showTemporaryView(request, temporary, text)
        }


        return showMainView(request, text)
    }

    private fun showMainView(request: CommandRequest, text: String): BotApiMethod<*> {
        val rooms = lightService.lightState.entries
            .associate { (room, state) ->
                val action = if (!state) "on" else "off"
                val currentStatus = if (state) " \uD83D\uDCA1" else ""
                Pair("$room/$action", getRusName(room) + currentStatus)
            }.toMutableMap()

        rooms["1min"] = "1 мин"
        rooms["5min"] = "5 мин"
        rooms["back"] = "\uD83D\uDD19 Назад"
        val buttons = createButtons(listOf(), rooms, 2)
        return createTmMessage(request, text, buttons)
    }

    private fun showTemporaryView(request: CommandRequest, temporary: Int, prefix: String): BotApiMethod<*> {

        val text = "$prefix\nВключить на $temporary мин:"

        val rooms = mutableMapOf<String, String>()
        lightService.lightState.forEach { (room: String, roomState: Boolean) ->
            val currentStatus = if (roomState) " \uD83D\uDCA1" else ""
            rooms[room] = getRusName(room) + ": " + currentStatus
        }
        rooms["back"] = "\uD83D\uDD19 Назад"
        var path = request.path
        if (path.size > 2) path = path.subList(0, 1)
        val buttons = createButtons(path, rooms, 2)
        return createTmMessage(request, text, buttons)
    }

    companion object {
        private const val LIGHT = "light"

        private fun getRusName(room: String) = if (room == "stairs-night") {
            "Ночник"
        } else {
            Room.getFromRoomCode(room).rusName
        }
    }
}
